src/main/java/com/google/cloud/tools/gradle/appengine/core/DownloadCloudSdkTask.java

```diff
@@ -37,35 +42,84 @@ public void setCloudSdkBuilderFactory(CloudSdkBuilderFactory cloudSdkBuilderFact
-        sdkVersion = "LATEST";
-      }
-      sdkHome = downloader.downloadSdk(sdkVersion);
-    } else {
-      if (!Strings.isNullOrEmpty(sdkVersion)) {
-        // Sdk home and version specified; validate installation
-        if (!downloader.isSdkValid(sdkVersion, sdkHome)) {
-          throw new GradleException(
-              "Specified Cloud SDK version and actual version of the SDK installed in the "
-                  + "specified directory do not match. You must either specify the correct "
-                  + "cloudSdkHome and cloudSdkVersion, or you can remove the cloudSdkHome field "
-                  + "to download the version you want.");
-        }
-      }
+    managedCloudSdk = managedCloudSdkFactory.newManagedSdk(sdkVersion);
+
+    ProgressListener progressListener =
+        new ProgressListener() {
+          @Override
+          public void start(String message, long totalWork) {
+            getLogger().lifecycle(message);
+          }
+
+          @Override
+          public void update(long workDone) {
+            // TODO: Show progress
+          }
+
+          @Override
+          public void update(String message) {
+            getLogger().lifecycle(message);
+          }
+
+          @Override
+          public void done() {}
+
+          @Override
+          public ProgressListener newChild(long allocation) {
+            return new ProgressListener() {
+              @Override
+              public void start(String message, long totalWork) {}
+
+              @Override
+              public void update(long workDone) {}
+
+              @Override
+              public void update(String message) {}
+
+              @Override
+              public void done() {}
+
+              @Override
+              public ProgressListener newChild(long allocation) {
+                return null;
+              }
+            };
+          }
+        };
+
+    ConsoleListener consoleListener =
+        new ConsoleListener() {
+          @Override
+          public void console(String rawString) {}
+        };
+
+    // Install sdk if not installed
+    if (!managedCloudSdk.isInstalled()) {
+      SdkInstaller installer = managedCloudSdk.newInstaller();
+      installer.install(progressListener, consoleListener);
+    }
+
+    // Install app engine component
+    if (!managedCloudSdk.hasComponent(SdkComponent.APP_ENGINE_JAVA)) {
+      SdkComponentInstaller componentInstaller = managedCloudSdk.newComponentInstaller();
+      componentInstaller.installComponent(
+          SdkComponent.APP_ENGINE_JAVA, progressListener, consoleListener);
+    }
+
+    // If version is set to LATEST, update Cloud SDK
+    if (!managedCloudSdk.isUpToDate()) {
+      SdkUpdater updater = managedCloudSdk.newUpdater();
+      updater.update(progressListener, consoleListener);
     }
 
-    cloudSdkBuilderFactory.setCloudSdkHome(sdkHome);
+    cloudSdkBuilderFactory.setCloudSdkHome(managedCloudSdk.getSdkHome().toFile());
```

Comment:
I didn't realize this earlier, but we probably don't want to be modifying extension properties from within a task execution. We might need to set this outside somewhere, probably in the AppEngineCorePluginConfiguration class.

And in that case, we're probably going to need to instantiate an instance of ManageCloudSdk out there too. You can file bugs to track this change, because it's probably going to mean that we don't pass a ManagedCloudSdkFactory in here, but just a ManagedCloudSdk instead. So maybe do it in a follow up PR.

Reply:
Filed bug #216